lib/smooth/core/io/UART.cpp

```diff
@@ -0,0 +1,574 @@
+    if(res != ESP_OK)
+    {
+      this->log_error("Failed to set rx full threshold", res);
+    }
+    return ESP_OK == res;
+  }
+
+  bool UART::set_tx_empty_threshold(int threshold)
+  {
+    const auto res = uart_set_tx_empty_threshold(this->port_, threshold);
+    if(res != ESP_OK)
+    {
+      this->log_error("Failed to set tx empty threshold", res);
+    }
+    return ESP_OK == res;
+  }
+
+  bool UART::set_rx_timeout(const uint8_t timeout)
+  {
+    const auto res = uart_set_rx_timeout(this->port_, timeout);
+    if(res != ESP_OK)
+    {
+      this->log_error("Failed to set rx timeout", res);
+    }
+    return ESP_OK == res;
+  }
+
+  bool UART::get_collision_flag(bool &flag)
+  {
+    const auto res = uart_get_collision_flag(this->port_, &flag);
+    if(res != ESP_OK)
+    {
+      this->log_error("Failed to get collision flag", res);
+    }
+    return ESP_OK == res;
+  }
+
+  bool UART::get_wakeup_threshold(int &threshold)
+  {
+    const auto res = uart_get_wakeup_threshold(this->port_, &threshold);
+    if(res != ESP_OK)
+    {
+      this->log_error("Failed to get wakeup threshold", res);
+    }
+    return ESP_OK == res;
+  }
+
+  bool UART::set_wakeup_threshold(int threshold)
+  {
+    const auto res = uart_set_wakeup_threshold(this->port_, threshold);
+    if(res != ESP_OK)
+    {
+      this->log_error("Failed to set wakeup threshold", res);
+    }
+    return ESP_OK == res;
+  }
+
+  bool UART::wait_tx_idle_polling()
+  {
+    const auto res = uart_wait_tx_idle_polling(this->port_);
+    if(res != ESP_OK)
+    {
+      this->log_error("Failed to wait for tx idle polling", res);
+    }
+    return ESP_OK == res;
+  }
+
+  bool UART::set_loop_back(bool enable)
+  {
+    const auto res = uart_set_loop_back(this->port_, enable);
+    if(res != ESP_OK)
+    {
+      this->log_error("Failed to set loop back", res);
+    }
+    return ESP_OK == res;
+  }
+
+  void UART::set_always_rx_timeout(bool enable)
+  {
+    uart_set_always_rx_timeout(this->port_, enable);
+  }
+
+  bool UART::driver_install(int rx_buffer_size, int tx_buffer_size, int queue_size, QueueHandle_t* uart_queue, int intr_alloc_flags)
+  {
+    const auto res = uart_driver_install(this->port_, rx_buffer_size, tx_buffer_size, queue_size, uart_queue, intr_alloc_flags);
+    if(res != ESP_OK)
+    {
+      this->log_error("Failed to install driver", res);
+    }
+    return ESP_OK == res;
+  }
+    
+  bool UART::driver_delete()
+  {
+    const auto res = uart_driver_delete(this->port_);
+    if(res != ESP_OK)
+    {
+      this->log_error("Failed to delete driver", res);
+    }
+    return ESP_OK == res;
+  }
+    
+  bool UART::is_driver_installed() const
+  {
+    return uart_is_driver_installed(this->port_);
+  }
+
+  void UART::log_error(const std::string msg, esp_err_t err)
+  {
+    std::stringstream ss;
+    ss << "UART Error for port '" << static_cast<int>(port_) << "', " << msg << ": " << esp_err_to_name(err);
```

Comment:
You never print out the error. There's also this: https://github.com/PerMalmberg/Smooth/blob/master/lib/smooth/include/smooth/core/logging/log.h